### ISSUE TITLE
[Core] Number of melee hits while dual wielding is incorrect

### ISSUE THREAD

**Issue by mercred** [dalaranwow/dalaran-wow#5677]
[//]: # (INFORMATION ABOUT TAGS)
[//]: # (The black and yellow tag is unique. There can only be one label assigned to them - one black and one yellow)
[//]: # (Priority - ... tags are assigned by team members)
[//]: # (Your issue will be reviewed by testers and the developers will fix it in the future.)
[//]: # (Only once the issue got the label [Fixed - On Live Server] you can find the fix on your Gameserver.)
[//]: # (You can increase the relevance of your issue by adding a thumb-up emote http://i.imgur.com/RFouVVi.png)

[//]: # (***************************************************************)
[//]: # (** !!!!!!!!!!!!!DON'T DELETE THIS TEMPLATE ELSE YOUR ISSUE WILL BE CLOSED!!!!!!!!!!!! **)
[//]: # (***************************************************************)

**Current behaviour**: Tell us what happens in detail (Where, When, What, Why, Who)
While dual wielding, the number of hits that should theoretically occur doesn't correspond to the actual number of hits. 
Actual number of hits **is off by 300+** in an 1hour test run,  which even if we factor slight miscalculations due to the omitted decimals is EXTREMELY fishy.  In other words, just **over 7%** of melee hits that should've occurred, simply didn't.

**Expected behaviour**: Tell us what should happen instead
The difference between theoretical and observed number of hits should be <10, with 10 being a VERY generous upper bound. After fixing this issue one should expect a difference of 1-2 hits.

**Steps to reproduce the problem**:

1.  Dualwield any weapons, make sure that you don't have any items equiped that might proc attack speed change, remove any talents that modify haste, reset recount/skada(you will need it for attack count).
2. Use this macro to start attacking:
/stopwatch play
/startattack
 and leave the character auto-attacking a dummy for one hour with occasional character jumps to prevent disconnect from the server
2. When stopwatch shows 60 minutes use the following macro to stop the test:
/stopwatch pause
/stopattack
3. Calculate theoretical number of hits:
As per https://wow.gamepedia.com/Attack_speed round to 4 significant digits.
No. of theoretical MH swings: MH_theory=test_time/MH_attack_speed;
No. of theoretical OH swings, same formula, as according to the Issue https://github.com/dalaranwow/dalaran-wow/issues/1060 the off-hand delay in the beginning was fixed: OH_theory=test_time/OH_attack_speed;
where test_time is overall duration of the test in seconds taken from stopwatch and MH_attack_speed/OH_attack_speed are taken from character info.
4. Summ MH_theory with OH_theory to get theoretical hit count and compare with an observed recount/skada counter.

**Include proofs for this behaviour**
Test 1. No items, no talents, 20min
![image](https://user-images.githubusercontent.com/15626206/51503363-e8f5a700-1dda-11e9-84b9-872f2212741e.png)
Theoretical number of MH attacks: 20*60/2,6=461,5385; OH=20*60/1,4=857,1429; Overall theoretical=1318,6813 with observed being 1255, resulting in a **missing 63** attacks or 5,0742%.
This test shows that there is an error in number of attacks even without any haste multipliers from talents/gear

Test 2. Items, no talents, 20min
![image](https://user-images.githubusercontent.com/15626206/51503260-55bc7180-1dda-11e9-89c5-bf9e08a59f3a.png)
Theoretical number of MH attacks: 20*60/2,24=535,7143; OH=20*60/1,29=930,2326; Overall theoretical=1465,9468 with observed being 1385, resulting in a **missing 80** attacks or 5,8445%.
This(and the next one) test demonstrates that percentage of hits that didn't occur changes when we apply additional haste multipliers, implemented properly it should stay the same. To be more specific in tests 3.1 and 3.2 we learn talent Lightning Reflexes https://wow.gamepedia.com/Lightning_Reflexes_(rogue_talent)  that gives us 10% haste.

Test 3.1. Items, talents(without hack and slash http://wowwiki.wikia.com/wiki/Hack_and_Slash), 20min
![image](https://user-images.githubusercontent.com/15626206/51503285-813f5c00-1dda-11e9-8bea-a2d58cc656d5.png)
Theoretical number of MH attacks: 20*60/2,04=588,2353; OH=20*60/1,17=1025,6410; Overall theoretical=1613,8763 with observed being 1507, resulting in a **missing 106** attacks or 7,09%.

Test 3.2 Items, talents( with hack and slash), 20min
![image](https://user-images.githubusercontent.com/15626206/51503304-9d42fd80-1dda-11e9-9681-2be8a36e8a33.png)
Theoretical number of MH attacks: 20*60/2,04=588,2353; OH=20*60/1,17=1025,6410; Overall theoretical=1613,8763 with observed being 1508, resulting in a **missing 105** attacks or 7,0210%.

Tests number 3.1 and 3.2 demonstrate that "Hack and slash"  talent has no influence on number of hits and is registered correctly. It also demonstrates bug reproducibility as we have a difference of only 1 missing attack in between these tests.

Test 4. Items, talents(with hack and slash), 60 minutes
![image](https://user-images.githubusercontent.com/15626206/51498497-0587e400-1dc7-11e9-9280-84750c745d10.png)
Theoretical number of MH attacks: 60*60/2,04=1764,7059; OH=60*60/1,17=3076,9231; Overall theoretical=4841,6290 with observed being 4506, resulting in a missing 335 attacks or 7,4485%.
Just a long 60 minute test to demostrate the bug using more data.

You can research this on several pages like http://wowhead.com or http://wowwiki.wikia.com or http://www.youtube.com/
Issues without proofs will be closed and tag On Hold will be added until the proof is provided.

**Include the ID for the game objects, npcs (creatures,pets,minions), spells, items, quests, instances, zones, achievements, skills**

You can research this on several pages like https://wowgaming.altervista.org/aowow/ or https://wowhead.com/

**Include Screenshots from the issue if necessary**

**Comment by alexuktc** [comment]
This sounds serious and could account for why rogues have lower dps than intended.

**Comment by mercred** [comment]
Any comments from devs on that? It seems like the more haste you have the more attacks you are missing, which potentially might be the reason assa.rogues have such a low dps being full raid buffed with very fast weapons.

**Comment by Lothloryen** [comment]
Hello there,

First of all, thank you very much for such a great report, it is really appreciated. I did take a look some time ago into this issue but couldn't find fix it properly at that moment, although I believe I managed to perform some improvements. Anyways, once I am done with some pending matters, I will work on this again with fresh eyes.

**Comment by donniDevito** [comment]
Possibly related to incorrect swing timer reset as it seems not to work properly https://github.com/dalaranwow/dalaran-wow/issues/4982
